src/functions/module_array.ts

```diff
@@ -0,0 +1,826 @@
+    assertSameDimensions(_lt("All the ranges must have the same dimensions."), ...args);
+    if (args.every(isMatrixArgValue)) {
+      let result = 0;
+      for (let i = 0; i < args[0].length; i++) {
+        for (let j = 0; j < args[0][i].length; j++) {
+          let product = 1;
+          for (const range of args) {
+            product *= toNumber(range[i][j]);
+          }
+          result += product;
+        }
+      }
+      return result;
+    }
+    return reduceNumbers(args, (acc, a) => acc * a, 1);
+  },
+  isExported: true,
+};
+
+// -----------------------------------------------------------------------------
+// SUMX2MY2
+// -----------------------------------------------------------------------------
+export const SUMX2MY2: AddFunctionDescription = {
+  description: _lt(
+    "Calculates the sum of the difference of the squares of the values in two array."
+  ),
+  args: [
+    arg(
+      "array_x (number, range<number>)",
+      _lt(
+        "The array or range of values whose squares will be reduced by the squares of corresponding entries in array_y and added together."
+      )
+    ),
+    arg(
+      "array_y (number, range<number>)",
+      _lt(
+        "The array or range of values whose squares will be subtracted from the squares of corresponding entries in array_x and added together."
+      )
+    ),
+  ],
+  returns: ["NUMBER"],
+  compute: function (arrayX: ArgValue, arrayY: ArgValue): number {
+    assertSameDimensions(
+      "The arguments array_x and array_y must have the same dimensions.",
+      arrayX,
+      arrayY
+    );
+
+    const _arrayX = flattenRowFirst([arrayX], toNumber);
+    const _arrayY = flattenRowFirst([arrayY], toNumber);
+
+    let result = 0;
+    for (let i = 0; i < _arrayX.length; i++) {
+      result += _arrayX[i] ** 2 - _arrayY[i] ** 2;
+    }
+
+    return result;
+  },
+  isExported: true,
+};
+
+// -----------------------------------------------------------------------------
+// SUMX2PY2
+// -----------------------------------------------------------------------------
+export const SUMX2PY2: AddFunctionDescription = {
+  description: _lt("Calculates the sum of the sum of the squares of the values in two array."),
+  args: [
+    arg(
+      "array_x (number, range<number>)",
+      _lt(
+        "The array or range of values whose squares will be added to the squares of corresponding entries in array_y and added together."
+      )
+    ),
+    arg(
+      "array_y (number, range<number>)",
+      _lt(
+        "The array or range of values whose squares will be added to the squares of corresponding entries in array_x and added together."
+      )
+    ),
+  ],
+  returns: ["NUMBER"],
+  compute: function (arrayX: ArgValue, arrayY: ArgValue): number {
+    assertSameDimensions(
+      "The arguments array_x and array_y must have the same dimensions.",
+      arrayX,
+      arrayY
+    );
+
+    const _array = flattenRowFirst([arrayX, arrayY], toNumber);
```

Comment:
It could be easier to read if we keep the same way to write it as in SUMX2MY2 and SUMXMY2 (flatten each separately and then a for loop), just to keep the same pattern ? Or it may be less effective ?

Reply:
The code would be a bit longer, but performances should ba about the same. But it's probably a good idea, `reduce` are always a bit hard to read.